netplugin/netd.go

```diff
@@ -14,248 +14,295 @@ limitations under the License.
-			log.Fatalf("Could not connect to kernel syslog")
-		}
-	} else {
-		u, err := url.Parse(syslogParam)
-		if err != nil {
-			log.Fatalf("Could not parse syslog spec: %v", err)
-		}
+	u, err := url.Parse(syslogParam)
+	if err != nil {
+		logrus.Fatalf("Could not parse syslog spec: %v", err)
+	}
 
-		hook, err = logrus_syslog.NewSyslogHook(u.Scheme, u.Host, syslog.LOG_INFO, "netplugin")
-		if err != nil {
-			log.Fatalf("Could not connect to syslog: %v", err)
-		}
+	hook, err = logrus_syslog.NewSyslogHook(u.Scheme, u.Host, syslog.LOG_INFO, "netplugin")
+	if err != nil {
+		logrus.Fatalf("Could not connect to syslog: %v", err)
 	}
 
-	log.AddHook(hook)
+	logrus.AddHook(hook)
 }
 
-func main() {
-	var opts cliOpts
-	var flagSet *flag.FlagSet
+func startNetPlugin(pluginConfig *plugin.Config) {
+	// Create a new agent
+	ag := agent.NewAgent(pluginConfig)
 
-	defHostLabel, err := os.Hostname()
+	// Process all current state
+	ag.ProcessCurrentState()
 
-	// parse rest of the args that require creating state
-	flagSet = flag.NewFlagSet("netplugin", flag.ExitOnError)
-	flagSet.BoolVar(&opts.debug,
-		"debug",
-		false,
-		"Show debugging information generated by netplugin")
-	flagSet.StringVar(&opts.syslog,
-		"syslog",
-		"",
-		"Log to syslog at proto://ip:port -- use 'kernel' to log via kernel syslog")
-	flagSet.BoolVar(&opts.jsonLog,
-		"json-log",
-		false,
-		"Format logs as JSON")
-	flagSet.StringVar(&opts.hostLabel,
-		"host-label",
-		defHostLabel,
-		"label used to identify endpoints homed for this host, default is host name. If -config flag is used then host-label must be specified in the the configuration passed.")
-	flagSet.StringVar(&opts.pluginMode,
-		"plugin-mode",
-		"docker",
-		"plugin mode docker|kubernetes|swarm-mode")
-	flagSet.StringVar(&opts.cfgFile,
-		"config",
-		"",
-		"plugin configuration. Use '-' to read configuration from stdin")
-	flagSet.StringVar(&opts.vtepIP,
-		"vtep-ip",
-		"",
-		"My VTEP ip address")
-	flagSet.StringVar(&opts.ctrlIP,
-		"ctrl-ip",
-		"",
-		"Local ip address to be used for control communication")
-	flagSet.Var(&opts.vlanIntf,
-		"vlan-if",
-		"VLAN uplink interface")
-	flagSet.BoolVar(&opts.version,
-		"version",
-		false,
-		"Show version")
-	flagSet.StringVar(&opts.dbURL,
-		"cluster-store",
-		"etcd://127.0.0.1:2379",
-		"state store url")
-	flagSet.StringVar(&opts.nwDriver,
-		"net-driver",
-		"ovs",
-		"network driver component ovs|vpp (vpp is not supported yet)")
-	flagSet.IntVar(&opts.vxlanUDPPort,
-		"vxlan-port",
-		4789,
-		"VxLAN UDP port number")
+	// post initialization processing
+	ag.PostInit()
 
-	err = flagSet.Parse(os.Args[1:])
-	if err != nil {
-		log.Fatalf("Failed to parse command. Error: %s", err)
+	// handle events
+	if err := ag.HandleEvents(); err != nil {
+		logrus.Infof("Netplugin exiting due to error: %v", err)
+		os.Exit(1)
 	}
+}
 
-	if opts.version {
-		fmt.Printf(version.String())
-		os.Exit(0)
-	}
+func validateConfigs(ctx *cli.Context) (plugin.Config, error) {
```

Comment:
this is not just validating, please choose another name (process, handle, etc.)

Reply:
which part is not doing validating ?

Reply:
validating means to check for correctness, it does not imply side effects, which this method is doing (this method validates & configures)

Reply:
it doesn't configure the plugin, it returns the configuration when it passed all validation, the plugin only created in `startNetPlugin`

Reply:
it configures syslog and logrus, maybe extract that part?